cpp/src/tsne/tsne.cu

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "common/cumlHandle.hpp"
+
+#include "tsne/tsne.h"
+
+#include "cublas_v2.h"
+#include "distances.h"
+#include "slow_kernels.h"
+#include "fast_kernels.h"
+#include "utils.h"
+#define cuda_max_potential cudaOccupancyMaxPotentialBlockSize
+
+
+namespace ML {
+using namespace MLCommon;
+
+void TSNE_fit(const cumlHandle &handle, const float *X, float *Y, const int n,
+            const int p, const int n_components = 2, int n_neighbors = 30,
+            const float perplexity = 30.0f, const int perplexity_max_iter = 100,
+            const int perplexity_tol = 1e-5,
+            const float early_exaggeration = 12.0f,
+            const int exaggeration_iter = 150, const float min_gain = 0.01f,
+            const float eta = 500.0f, const int max_iter = 500,
+            const float pre_momentum = 0.8, const float post_momentum = 0.5,
+            const long long seed = -1, const bool initialize_embeddings = true,
+            const bool verbose = true, const int method = 1)
+    // Method = 0 for Naive, 1 for Fast
+{
+    assert(n > 0 && p > 0 && n_components > 0 && n_neighbors > 0 && X != NULL && Y != NULL);
+    if (verbose)
+        printf("[Info]  Data size = (%d, %d) with n_components = %d\n", n, p, n_components);
+
+    auto d_alloc = handle.getDeviceAllocator();
+    cudaStream_t stream = handle.getStream();
+
+    if (n_neighbors > n) n_neighbors = n;
+
+    // Some preliminary intializations for cuBLAS and cuML
+    const int k = n_components;
+    cublasHandle_t BLAS = handle.getImpl().getCublasHandle();
```

Comment:
It looks like your current version only uses this for the `naive` method. Move this into that block to make this more readable.